modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11y.tsx

```diff
@@ -0,0 +1,193 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import './A11y.scss';
+
+import ClayIcon from '@clayui/icon';
+import ClayList from '@clayui/list';
+import ClayPopover from '@clayui/popover';
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+import ReactDOM from 'react-dom';
+
+import useA11y from '../../hooks/useA11y';
+
+import type {Violation as TViolation} from '../../hooks/useA11y';
+import type {A11yCheckerOptions} from './A11yChecker';
+
+const rectAttrs: Array<keyof DOMRect> = [
+	'bottom',
+	'height',
+	'left',
+	'right',
+	'top',
+	'width',
+];
+
+interface IRectState {
+	rect: DOMRect | undefined;
+	hasRectChanged: boolean;
+}
+
+const rectChanged = (a: DOMRect = {} as DOMRect, b: DOMRect = {} as DOMRect) =>
+	rectAttrs.some((prop) => a[prop] !== b[prop]);
+
+const useObserveRect = (
+	callback: (rect: DOMRect | undefined) => void,
+	node: Element | null
+) => {
+	const rafIdRef = useRef<number>();
+
+	const run = useCallback(
+		(node: Element, state: IRectState) => {
+			const newRect = node.getBoundingClientRect();
+
+			if (rectChanged(newRect, state.rect)) {
+				state.rect = newRect;
+
+				callback(state.rect);
+			}
+
+			rafIdRef.current = window.requestAnimationFrame(() =>
+				run(node, state)
+			);
+		},
+		[callback]
+	);
+
+	useEffect(() => {
+		if (node) {
+			run(node, {
+				hasRectChanged: false,
+				rect: undefined,
+			});
+
+			return () => {
+				if (rafIdRef.current) {
+					cancelAnimationFrame(rafIdRef.current);
+				}
+			};
+		}
+	}, [node, run]);
+};
+
+const Overlay = React.forwardRef<
+	HTMLDivElement,
+	React.ButtonHTMLAttributes<HTMLDivElement>
+>(({style, ...othersProps}, ref) =>
+	ReactDOM.createPortal(
+		<div {...othersProps} className="a11y-overlay" ref={ref} style={style}>
+			<div className="a11y-indicator">
+				<ClayIcon symbol="info-circle" />
+			</div>
+			<div className="a11y-backdrop" />
+		</div>,
+		document.body
+	)
+);
+
+type ViolationProps = {
+	modifyIndex: number;
+	target: string;
+	violations: Array<TViolation>;
+};
```

Comment:
I don't think we actually have a solid rule for this, but in Clay we typically use `interface` for declaring props. Do you recall any previous conversation where we decided type vs interface? I don't have strong opinions for one or the other, but I can imagine future devs might be unsure which one we prefer

Reply:
I like the idea of "Use Interface until You Need Type" since they are a bit more open. This way when someone uses a `type`, we can be sure it's intentional. But maybe our rule of thumb is to just point to the cheatsheet and handbook and not worry too much about it.

Reply:
Yeah I like that too, I think we could start a conversation in the guidelines to follow some cheatsheets, sometimes I get confused in my mind as to whether I should follow the Clay standard for interface and type nomenclature or not.